test/SourceKit/CursorInfo/cursor_symbol_graph_objc.swift

```diff
@@ -1,92 +1,268 @@
+// CHECK-FUNC:             "spelling": "func"
+// CHECK-FUNC:           },
+// CHECK-FUNC:           {
+// CHECK-FUNC:             "kind": "text",
+// CHECK-FUNC:             "spelling": " "
+// CHECK-FUNC:           },
+// CHECK-FUNC:           {
+// CHECK-FUNC:             "kind": "identifier",
+// CHECK-FUNC:             "spelling": "someFunc"
+// CHECK-FUNC:           },
+// CHECK-FUNC:           {
+// CHECK-FUNC:             "kind": "text",
+// CHECK-FUNC:             "spelling": "()"
+// CHECK-FUNC:           }
+// CHECK-FUNC:         ],
+// CHECK-FUNC:         "title": "someFunc()"
+// CHECK-FUNC:       },
+// CHECK-FUNC:       "pathComponents": [
+// CHECK-FUNC:         "someFunc()"
+// CHECK-FUNC:       ]
+// CHECK-FUNC:     }
+// CHECK-FUNC:   ]
+// CHECK-FUNC: }
+// CHECK-FUNC: SYMBOL GRAPH END
+
+//--- mod/module.modulemap
+module MyMod {
+  header "M.h"
+}
+
+//--- mod/M.h
+/// someFunc doc
+void someFunc(void);
+
+struct ObjCStruct {
+  int noDoc;
+  // CHECK-NO-NOT: "docComment"
+
+  // regular comment
+  int regComments;
+  // CHECK-REG-NOT: "docComment"
+
+  /// single line doc
+  int simpleSingle1;
+  // CHECK-SINGLE1: "docComment": {
+  // CHECK-SINGLE1-NEXT:    "lines": [
+  // CHECK-SINGLE1-NEXT:      {
+  // CHECK-SINGLE1-NEXT:        "text": "single line doc"
+  // CHECK-SINGLE1-NEXT:      }
+  // CHECK-SINGLE1-NEXT:    ]
+  // CHECK-SINGLE1-NEXT:  }
+
+  //! single line doc
+  int simpleSingle2;
+  // CHECK-SINGLE2: "docComment": {
+  // CHECK-SINGLE2-NEXT:    "lines": [
+  // CHECK-SINGLE2-NEXT:      {
+  // CHECK-SINGLE2-NEXT:        "text": "single line doc"
+  // CHECK-SINGLE2-NEXT:      }
+  // CHECK-SINGLE2-NEXT:    ]
+  // CHECK-SINGLE2-NEXT:  }
+
+  /** single line block doc */
+  int simpleBlock1;
+  // CHECK-BLOCK1: "docComment": {
+  // CHECK-BLOCK1-NEXT:    "lines": [
+  // CHECK-BLOCK1-NEXT:      {
+  // CHECK-BLOCK1-NEXT:        "text": "single line block doc "
+  // CHECK-BLOCK1-NEXT:      }
+  // CHECK-BLOCK1-NEXT:    ]
+  // CHECK-BLOCK1-NEXT:  }
+
+  /*! single line block doc */
+  int simpleBlock2;
+  // CHECK-BLOCK2: "docComment": {
+  // CHECK-BLOCK2-NEXT:    "lines": [
+  // CHECK-BLOCK2-NEXT:      {
+  // CHECK-BLOCK2-NEXT:        "text": "single line block doc "
+  // CHECK-BLOCK2-NEXT:      }
+  // CHECK-BLOCK2-NEXT:    ]
+  // CHECK-BLOCK2-NEXT:  }
+
+  /**
+   * docs with art prefix
+   *   indent
+   * last
+   */
+  int artBlock;
+  // CHECK-ART: "docComment": {
+  // CHECK-ART-NEXT:    "lines": [
+  // CHECK-ART-NEXT:      {
+  // CHECK-ART-NEXT:        "text": ""
+  // CHECK-ART-NEXT:      },
+  // CHECK-ART-NEXT:      {
+  // CHECK-ART-NEXT:        "text": " docs with art prefix"
+  // CHECK-ART-NEXT:      },
+  // CHECK-ART-NEXT:      {
+  // CHECK-ART-NEXT:        "text": "   indent"
+  // CHECK-ART-NEXT:      },
+  // CHECK-ART-NEXT:      {
+  // CHECK-ART-NEXT:        "text": " last"
+  // CHECK-ART-NEXT:      },
+  // CHECK-ART-NEXT:      {
+  // CHECK-ART-NEXT:        "text": "   "
```

Comment:
The awkward newline and whitespace handling surrounding block doc comments makes me a little concerned. While this can be easily handled in a Markdown parser after the fact, the raw text coming out of Clang here is a bit confusing, especially this last line here. It's probably not something we should fix in this PR, but part of me would like to look into it at some point.

Reply:
Oh definitely, I mentioned that in the commit message but maybe that wasn't clear enough:
```
There's also a few odd
cases with mixed doc types and when each line is prefixed with '*', but
it's good enough for an initial implementation.
```

Mostly I was hoping that the new `getFormattedLines` handles these cases properly already, so I didn't look into it too much. We should use that when we can + fix it if it still has these weird cases.

Reply:
Thanks for the clarification! I'm not aware of `getFormattedLines`, but if it's something we can use to fix this in the future, then i'm all for it.

Reply:
See https://reviews.llvm.org/D119479

Reply:
Actually I did try something in the new `getFormattedLines` to rectify this 😅
See https://github.com/llvm/llvm-project/blob/e9c8d0ff71baba2702765d6cd173421d54883fee/clang/lib/AST/RawCommentList.cpp#L400-L408 and additional logics with `PreviousLine` in that method.
I just gave it a try and apparently it didn't fully fix the excessive leading/trailing whitespace issue... The only diff of output using `getFormattedLines` is that there's only one space character in the last line, instead of two here.
It would definitely require more work to fix this as the lexer handles these block comments a bit weird IMO. On top of my head I think just a simple post-process to remove any leading/trailing whitespace lines would work great and easy.

Reply:
Oh cool, thanks @zixu-w! Seems like that would handle most of these. There's still the ASCII art blocks, where they contain the initial space. Naively we could include the star+whitespace in the calculation for the indent level, but then that gets tricky with things like:
```
/** Not sure if we care about this case, but here's some dot points:
   * this
   * is actually
   * a list of dot point
*/
```

Pretty odd though, so 🤷